File: cyanBat/src/main/java/at/msmiech/cyanbat/objects/gameobjects/CyanBat.java
package at.msmiech.cyanbat.objects.gameobjects;

import java.util.ArrayList;
import java.util.List;

import android.graphics.Rect;
import android.util.Log;
import at.grueneis.game.framework.Input.TouchEvent;
import at.grueneis.game.framework.Graphics;
import at.grueneis.game.framework.Pixmap;
import at.msmiech.cyanbat.CyanBatGame;
import at.msmiech.cyanbat.interfaces.Collidable;
import at.msmiech.cyanbat.interfaces.GameObject;
import at.msmiech.cyanbat.screens.CyanBatBaseScreen;
import at.msmiech.cyanbat.screens.FailScreen;
import at.msmiech.cyanbat.screens.GameScreen;

public class CyanBat extends PixmapGameObject implements Collidable {

	private float ANIM_TICK_INTERVAL = 0.2f;
	private final float TICK_INITIAL = 0.009f;
	private float animTickTime = 0;
	public boolean alive = true;
	public int lives = 3;
	private int animTick;
	private float tickTime = 0;
	private float tick = TICK_INITIAL;
	private int srcX;
	private GameScreen gs;
	private ArrayList<CyanCurvature> curvatures = new ArrayList<CyanCurvature>();
	public final static int realWidth = 45;

	public CyanBat(int x, int y, int width, int height, Pixmap pm, GameScreen gs) {
		super(new Rect(x, y, x + realWidth, y + height), pm);
		this.gs = gs;
		init();
	}

	private void init() {
		// TODO Auto-generated method stub

	}

	@Override
	public void update(float deltaTime, List<TouchEvent> touchEvents) {
		if (DEBUG)
			Log.d(TAG, "updateBat");
		updateLogic(deltaTime, touchEvents);
		updateAnimation(deltaTime);
		updateCurvatations();

		if (CyanBatGame.SHOOTING_ENABLED) {
			if (CyanBatBaseScreen.game.getInput().getTouchEvents().size() > 1)
				shoot();
		}
		super.update(deltaTime, touchEvents);
	}

	private void updateCurvatations() {
		Rect potentialRect = new Rect(rect.left,
				(rect.top + (rect.height() / 2)), rect.left + rect.width() / 4,
				rect.bottom);
		for (int i = 0; i < curvatures.size(); i++) {
			CyanCurvature curv = curvatures.get(i);
			if (curv == null || curv.removeMe) {
				curvatures.remove(curv);
				continue;
			}
			if (Rect.intersects(curv.rect, potentialRect)) {
				return;
			}
		}

		potentialRect.left -= 2;
		potentialRect.right -= 1;
		CyanCurvature go = new CyanCurvature(potentialRect);
		curvatures.add(go);
		gs.gameObjects.add(go);
	}

	private void shoot() {
		if (Shot.count > 1)
			return;
		AbstractGameObject shot = new Shot(new Rect(rect.left, rect.top,
				rect.left + CyanBatGame.shot.getWidth(), rect.top
						+ CyanBatGame.shot.getHeight()), CyanBatGame.shot, this);
		gs.gameObjects.add((GameObject) shot);
		gs.colChk.addObjectToCheck(shot);
	}

	private void updateAnimation(float deltaTime) {
		animTickTime += deltaTime;
		if (animTickTime > ANIM_TICK_INTERVAL) {
			animTick = (animTick + 1) % 2;
			animTickTime -= ANIM_TICK_INTERVAL;
		}
		switch (animTick) {
		case 0:
			srcX = 0;
			break;
		case 1:
			srcX = 50;
			break;
		}
	}

	private void updateLogic(float deltaTime, List<TouchEvent> touchEvents) {
		if (alive) {
			velocity.x = 0;
			velocity.y = 0;
			if (!touchEvents.isEmpty()) {
				tickTime += deltaTime;
				while (tickTime > tick) {
					tickTime -= tick;
					for (int i = 0; i < touchEvents.size(); i++) {
						TouchEvent touch = touchEvents.get(i);
						if (touch.type == TouchEvent.TOUCH_DRAGGED)
							move(touch);
					}

				}
			}
		} else {
			velocity.x = 0;
			velocity.y = 2;
			if (rect.top > CyanBatBaseScreen.DISPLAY_WIDTH)
				CyanBatBaseScreen.game
						.setScreen(new FailScreen(GameScreen.game));
		}
	}

	private void move(TouchEvent touch) {
		if (DEBUG)
			Log.d(TAG, "moveBat");
		if (touch.x > rect.left) {
			if (rect.right < CyanBatBaseScreen.DISPLAY_HEIGHT) {
				velocity.x = 3;
			}
		} else {
			if (rect.left > 0) {
				velocity.x = -3;
			}
		}
		if (touch.y > rect.bottom) {
			if (rect.bottom < CyanBatBaseScreen.DISPLAY_WIDTH) {
				velocity.y = 3;
			}
		} else {
			if (rect.top > 0) {
				velocity.y = -3;
			}
		}

	}

	@Override
	public void draw(Graphics g) {
		if (DEBUG)
			Log.d(TAG, "drawBat");
		g.drawPixmap(pixmap, rect.left, rect.top, srcX, 0, realWidth,
				rect.height());
		super.draw(g);
	}

	public void hit() {
		CyanBatGame.vib.vibrate(250);
		lives -= 1;
		if (lives < 1) {
			CyanBatGame.deathSound.play(100);
			alive = false;
			gs.saveHighscore();
			CyanBatGame.musicPlayer.stopMusic();
			gs.interruptThreads();
			CyanBatGame.gameOverMusic.play();
		}
	}
}

File: cyanBat/src/main/java/at/smiech/cyanbat/gameobject/impl/CyanBat.kt
package at.smiech.cyanbat.gameobject.impl

import android.graphics.Rect
import android.os.VibrationEffect
import android.os.VibrationEffect.DEFAULT_AMPLITUDE
import android.util.Log
import at.grueneis.game.framework.Graphics
import at.grueneis.game.framework.Input.TouchEvent
import at.grueneis.game.framework.Pixmap
import at.smiech.cyanbat.activity.CyanBatGameActivity
import at.smiech.cyanbat.gameobject.Collidable
import at.smiech.cyanbat.gameobject.PixmapGameObject
import at.smiech.cyanbat.ui.game.GameOverScreen
import at.smiech.cyanbat.ui.game.GameScreen
import at.smiech.cyanbat.util.DEBUG
import at.smiech.cyanbat.util.TAG
import java.util.*

class CyanBat(
    x: Int,
    y: Int,
    width: Int = DEFAULT_WIDTH,
    height: Int,
    pixmap: Pixmap,
    private val frameBufferWidth: Int,
    private val frameBufferHeight: Int,
    private val gameScreen: GameScreen
) : PixmapGameObject(Rect(x, y, x + width, y + height), pixmap), Collidable {

    private var animTickTime = 0f
    var alive = true
    var lives = 3
    private var animTick: Int = 0
    private var tickTime = 0f
    private val tick = TICK_INITIAL
    private var srcX: Int = 0
    private val trails = ArrayList<CyanTrail>()
    private var hitCooldown = MAX_HIT_COOLDOWN // cooldown grace period in seconds

    override fun update(deltaTime: Float, touchEvents: List<TouchEvent>) {
        if (DEBUG)
            Log.d(TAG, "updateBat")
        updateLogic(deltaTime, touchEvents)
        updateAnimation(deltaTime)
        updateTrails()

        super.update(deltaTime, touchEvents)
    }

    private fun updateTrails() {
        val potentialRect = Rect(
            rectangle.left,
            rectangle.top + rectangle.height() / 2, rectangle.left + rectangle.width() / 4,
            rectangle.bottom
        )
        if (trails.any { Rect.intersects(it.rectangle, potentialRect) }) {
            return
        }

        trails.removeIf { curve -> curve.removeMe }

        potentialRect.left -= 2
        potentialRect.right -= 1
        val go = CyanTrail(potentialRect)
        trails.add(go)
        gameScreen.gameObjects.add(go)
    }

    private fun shoot() {
        if (Shot.count > 1)
            return
        val shot = Shot(
            Rect(
                rectangle.left,
                rectangle.top,
                rectangle.left + CyanBatGameActivity.gameAssets.graphics.shot.width,
                rectangle.top + CyanBatGameActivity.gameAssets.graphics.shot.height
            ), CyanBatGameActivity.gameAssets.graphics.shot, this
        )
        gameScreen.gameObjects.add(shot)
        gameScreen.colChk.addObjectToCheck(shot)
    }

    private fun updateAnimation(deltaTime: Float) {
        animTickTime += deltaTime
        if (animTickTime > ANIM_TICK_INTERVAL) {
            animTick = (animTick + 1) % 2
            animTickTime -= ANIM_TICK_INTERVAL
        }
        when (animTick) {
            0 -> srcX = 0
            1 -> srcX = 50
        }
    }

    private fun updateLogic(deltaTime: Float, touchEvents: List<TouchEvent>) {
        if (alive) {
            if (hitCooldown > 0f) {
                hitCooldown -= deltaTime
            }
            velocity.x = 0f
            velocity.y = 0f
            if (touchEvents.isNotEmpty()) {
                tickTime += deltaTime
                while (tickTime > tick) {
                    tickTime -= tick
                    for (i in touchEvents.indices) {
                        val touch = touchEvents[i]
                        if (touch.type == TouchEvent.TOUCH_DRAGGED)
                            move(touch)
                    }

                }
            }
        } else {
            velocity.x = 0f
            velocity.y = 2f
            // check whether the bat animation has finished (i.e. bat is outside the screen)
            if (rectangle.top > frameBufferWidth) {
                // switch to the GameOverScreen, if so
                gameScreen.game.setScreen(GameOverScreen(gameScreen.game))
            }
        }
    }

    private fun move(touch: TouchEvent) {
        if (DEBUG)
        {
            Log.d(TAG, "moveBat")
        }
        if (touch.x > rectangle.centerX()) {
            if (rectangle.right < frameBufferWidth) {
                velocity.x = 3f
            }
        } else {
            if (rectangle.left > 0) {
                velocity.x = -3f
            }
        }
        if (touch.y > rectangle.centerY()) {
            if (rectangle.bottom < frameBufferHeight) {
                velocity.y = 3f
            }
        } else {
            if (rectangle.top > 0) {
                velocity.y = -3f
            }
        }

    }

    override fun draw(g: Graphics) {
        if (DEBUG)
            Log.d(TAG, "drawBat")
        g.drawPixmap(
            pixmap, rectangle.left, rectangle.top, srcX, 0, DEFAULT_WIDTH,
            rectangle.height()
        )
        super.draw(g)
    }

    override fun hit() {
        CyanBatGameActivity.gameAssets.vib.vibrate(VibrationEffect.createOneShot(250, DEFAULT_AMPLITUDE))
        if (hitCooldown <= 0.01f) {
            lives -= 1
            hitCooldown = MAX_HIT_COOLDOWN
        }
        if (lives < 1) {
            lives = 0
            if (CyanBatGameActivity.soundsEnabled) {
                CyanBatGameActivity.gameAssets.audio.deathSound.play(100f)
            }
            alive = false
            gameScreen.saveHighscore()
            CyanBatGameActivity.gameAssets.audio.gameTrack.apply {
                stop()
                isLooping = false
            }
            if (CyanBatGameActivity.musicEnabled) {
                CyanBatGameActivity.gameAssets.audio.gameOverMusic.play()
            }
        }
    }

    companion object {
        const val DEFAULT_WIDTH = 45
        private const val ANIM_TICK_INTERVAL = 0.2f
        private const val TICK_INITIAL = 0.009f
        private const val MAX_HIT_COOLDOWN = 0.5f
    }
}
